src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -0,0 +1,1459 @@
+        private void Emit()
+        {
+            // emit all values on the stack
+            while (_valueStack.Count > 0)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+                var value = _valueStack.Pop();
+                WriteVariant(value);
+            }
+        }
+
+        private void WriteVariant(Variant value)
+        {
+            switch (value.Kind)
+            {
+                case VariantKind.Null:
+                    _writer.Write((byte)EncodingKind.Null);
+                    break;
+
+                case VariantKind.Boolean:
+                    _writer.Write((byte)(value.AsBoolean() ? EncodingKind.Boolean_T : EncodingKind.Boolean_F));
+                    break;
+
+                case VariantKind.Byte:
+                    _writer.Write((byte)EncodingKind.UInt8);
+                    _writer.Write(value.AsByte());
+                    break;
+
+                case VariantKind.SByte:
+                    _writer.Write((byte)EncodingKind.Int8);
+                    _writer.Write(value.AsSByte());
+                    break;
+
+                case VariantKind.Int16:
+                    _writer.Write((byte)EncodingKind.Int16);
+                    _writer.Write(value.AsInt16());
+                    break;
+
+                case VariantKind.UInt16:
+                    _writer.Write((byte)EncodingKind.UInt16);
+                    _writer.Write(value.AsUInt16());
+                    break;
+
+                case VariantKind.Int32:
+                    {
+                        var v = value.AsInt32();
+                        if (v >= 0 && v <= 10)
+                        {
+                            _writer.Write((byte)((int)EncodingKind.Int32_0 + v));
+                        }
+                        else if (v >= 0 && v < byte.MaxValue)
+                        {
+                            _writer.Write((byte)EncodingKind.Int32_B);
+                            _writer.Write((byte)v);
+                        }
+                        else if (v >= 0 && v < ushort.MaxValue)
+                        {
+                            _writer.Write((byte)EncodingKind.Int32_S);
+                            _writer.Write((ushort)v);
+                        }
+                        else
+                        {
+                            _writer.Write((byte)EncodingKind.Int32);
+                            _writer.Write(v);
+                        }
+                    }
+                    break;
+
+                case VariantKind.UInt32:
+                    {
+                        var v = value.AsUInt32();
+                        if (v >= 0 && v <= 10)
+                        {
+                            _writer.Write((byte)((int)EncodingKind.UInt32_0 + v));
+                        }
+                        else if (v >= 0 && v < byte.MaxValue)
+                        {
+                            _writer.Write((byte)EncodingKind.UInt32_B);
```

Comment:
There is no such requirement being expressed here.  This math is only determining if a particular value will fit within an encoding, not that no other encoding can encode this value.